sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+      bit_index = SPI_DEVICE_INTR_TEST_RXF;
+      break;
+    case kDifSpiDeviceIrqTypeRxAboveLevel:
+      bit_index = SPI_DEVICE_INTR_TEST_RXLVL;
+      break;
+    case kDifSpiDeviceIrqTypeTxBelowLevel:
+      bit_index = SPI_DEVICE_INTR_TEST_TXLVL;
+      break;
+    case kDifSpiDeviceIrqTypeRxError:
+      bit_index = SPI_DEVICE_INTR_TEST_RXERR;
+      break;
+    case kDifSpiDeviceIrqTypeRxOverflow:
+      bit_index = SPI_DEVICE_INTR_TEST_RXOVERFLOW;
+      break;
+    case kDifSpiDeviceIrqTypeTxUnderflow:
+      bit_index = SPI_DEVICE_INTR_TEST_TXUNDERFLOW;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  // TODO: Add a variant of set_bit for writeonly registers?
+  uint32_t mask = 1 << bit_index;
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_INTR_TEST_REG_OFFSET, mask);
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_set_irq_levels(
+    const dif_spi_device_t *spi, uint16_t rx_level, uint16_t tx_level) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t compressed_limit = 0;
+  compressed_limit |= (rx_level & SPI_DEVICE_FIFO_LEVEL_RXLVL_MASK)
+                      << SPI_DEVICE_FIFO_LEVEL_RXLVL_OFFSET;
+  compressed_limit |= (tx_level & SPI_DEVICE_FIFO_LEVEL_TXLVL_MASK)
+                      << SPI_DEVICE_FIFO_LEVEL_TXLVL_OFFSET;
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_FIFO_LEVEL_REG_OFFSET,
+                      compressed_limit);
+
+  return kDifSpiDeviceResultOk;
+}
+
+/**
+ * Parameters for compressing and decompressing a FIFO pointer register.
+ */
+typedef struct fifo_ptr_params {
+  ptrdiff_t reg_offset;
+  ptrdiff_t write_offset;
+  ptrdiff_t read_offset;
+  uint32_t write_mask;
+  uint32_t read_mask;
+} fifo_ptr_params_t;
+
+/**
+ * Parameters for the transmission FIFO.
+ */
+static const fifo_ptr_params_t kTxFifoParams = {
+    .reg_offset = SPI_DEVICE_TXF_PTR_REG_OFFSET,
+    .write_offset = SPI_DEVICE_TXF_PTR_WPTR_OFFSET,
+    .write_mask = SPI_DEVICE_TXF_PTR_WPTR_MASK,
+    .read_offset = SPI_DEVICE_TXF_PTR_RPTR_OFFSET,
+    .read_mask = SPI_DEVICE_TXF_PTR_RPTR_MASK,
+};
+
+/**
+ * Parameters for the receipt FIFO.
+ */
+static const fifo_ptr_params_t kRxFifoParams = {
+    .reg_offset = SPI_DEVICE_RXF_PTR_REG_OFFSET,
+    .write_offset = SPI_DEVICE_RXF_PTR_WPTR_OFFSET,
+    .write_mask = SPI_DEVICE_RXF_PTR_WPTR_MASK,
+    .read_offset = SPI_DEVICE_RXF_PTR_RPTR_OFFSET,
+    .read_mask = SPI_DEVICE_RXF_PTR_RPTR_MASK,
+};
+
+/**
+ * An exploded FIFO pointer value, consisting of a `uint11_t`
+ * offset part (an offset into a FIFO), and a `uint1_t` phase
+ * (which indicates whether this pointer has wrapped around or not).
+ *
+ * See also `fifo_ptrs_t`.
+ */
+typedef struct fifo_ptr {
+  uint16_t offset;
+  bool phase;
+} fifo_ptr_t;
+
+// Masks for extracting the phase and offset parts from a
+// compressed FIFO pointer.
+static const uint16_t kPhaseMask = (1 << 11);
+static const uint16_t kOffsetMask = (1 << 11) - 1;
```

Comment:
Names could be a little more specific, e.g. `kFifoPhaseMask` and `kFifoOffsetMask`

Reply:
Done.